File: from-1501-to-1600/1552-magnetic-force-between-two-balls/solution.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        n = len(position)
        position.sort()

        def countballs(dist: int) -> int:
            num_balls = 1
            curr = position[0]
            for pos in position[1:]:
                if pos - curr >= dist:
                    num_balls += 1
                    curr = pos
            return num_balls

        left, right = 1, position[-1] - position[0]
        while left <= right:
            mid = (left + right) // 2
            balls = countballs(mid)
            if balls >= m:
                left = mid + 1
            else:
                right = mid - 1

        return right 

File: from-1501-to-1600/1552-magnetic-force-between-two-balls/solution.go
func maxDistance(position []int, m int) int {
    n := len(position)
    sort.Ints(position)

    countBalls := func(dist int) int {
        numBalls := 1
        curr := position[0]
        for _, pos := range position[1:] {
            if pos - curr >= dist {
                numBalls += 1
                curr = pos
            }
        }
        return numBalls
    }

    left, right := 1, position[n - 1] - position[0]
    for left <= right {
        mid := left + (right - left) / 2
        if countBalls(mid) >= m {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    return right
}